README.md
# jQuery RadiantScroller plugin

This is a jQuery plugin that allows to create responsive scrollers (carousels) with grid and simple horizontal layouts.
RadiantScroller can be cuztomized with the variety of options. [API methods](#api) and [callbacks](#callbacks) are
available. Elements may also have [animated captions](#captions).

The current version is 0.1.0 (29/07/2014).

Documentation and demos can also be found at
[http://www.radiant-wind.com/plugins/radiant_scroller](http://www.radiant-wind.com/plugins/radiant_scroller).

## Demos

Images were taken from [FreeImages](http://www.freeimages.co.uk).

* [Basic scroller with 3x2 grid with no mousewheel support and no pagination](http://www.radiant-wind.com/code_examples/10)
* [Scroller with 3x2 grid with mousewheel support and pagination](http://www.radiant-wind.com/code_examples/11)
* [Scroller with one column and two rows](http://www.radiant-wind.com/code_examples/13)
* [Basic Radiant Scroller's API](http://www.radiant-wind.com/code_examples/12) (demonstrates that you can scroll by
specified number of pages or directly to the page with specified number)
* [Radiant Scroller with captions](http://www.radiant-wind.com/code_examples/18) (also demonstrates how to add
jQueryUI's easings)
* [Radiant Scroller's callback API](http://www.radiant-wind.com/code_examples/19)

## Requirements

1. jQuery 1.7.0+
2. [MouseWheel plugin](https://github.com/brandonaaron/jquery-mousewheel) (only if you want to provide mousewheel support)

## Installation and usage

* Hook up required files

* Hook up `jquery.radiant_scroller.min.js` and `jquery.radiant_scroller.css`

* Set up basic layout for the scroller:

```html
<div id="your_scroller_id">
    <div class="scroller-el"><img src="image1.jpg" alt="Image1" /></div>
    <div class="scroller-el"><img src="image2.jpg" alt="Image2" /></div>
    <div class="scroller-el"><img src="image3.jpg" alt="Image3" /></div>
</div>
```

The `scroller-el` class is required.

* Set up basic styling. In fact, only specifying margins and dimensions for the elements in the scroller is required. This can be done like:

```css
.radiant_scroller .scroller-el {
    width: 200px;
    height: 200px;
    margin-right: 10px;
    margin-bottom: 10px; /* This is optional but recommended - otherwise your images will have no horizontal spacing */
}
```

* Initialize the scroller. The values for `elementWidth` and `elementMargin` should be the same as your specified in
the previous step otherwise the scroller's grid will be built incorrectly.
You will probably want to specify scroller's grid dimensions with the `cols` and `rows` attributes (the default is
2x2 grid). All available options are listed in the "Options" section.

```js
$(document).ready(function() {
    $('#myScroller').radiantScroller({
        elementWidth: 200,
        elementMargin: 10,
        cols: 3,
        rows: 3
    });
});
```

* Style your scroller as needed (check [Styling section](#styling) and [demos](#demos) for more information).

## Options

<table>
  <thead>
    <tr>
      <th>Property</th>
      <th>Default</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th scope="row"><code>animateDuration</code></th>
      <td><code>700</code></td>
      <td><code>Integer</code> Animation duration for the scrolling. The value is specified in milliseconds.</td>
    </tr>
    <tr>
      <th scope="row"><code>addPagination</code></th>
      <td><code>false</code></td>
      <td><code>Boolean</code> Whether the scroller should have pagination
      (often displayed as small navigational dots) enabled.</td>
    </tr>
    <tr>
      <th scope="row"><code>captionsAnimateDuration</code></th>
      <td><code>400</code></td>
      <td><code>Integer</code> Animation duration for the captions.</td>
    </tr>
    <tr>
      <th scope="row"><code>captionsAnimateEasingType</code></th>
      <td><code>'swing'</code></td>
      <td><code>Integer</code> Easing type for the captions' animation. You can specify any other type of easing but
      bear in mind
      that jQuery has only <code>swing</code> and <code>linear</code> easings included.
      You will have to include <a target="_blank" href="http://jqueryui.com/effect/#easing">jQueryUI's effects and
      easings plugin</a> to get more.</td>
    </tr>
    <tr>
      <th scope="row"><code>cols</code></th>
      <td><code>2</code></td>
      <td><code>Integer</code> How many (maximum) columns should the scroller have - basically this means how many elements
      will be visible at once horizontally (but if the scroller's width changes this value will also change).</td>
    </tr>
    <tr>
      <th scope="row"><code>delayDuration</code></th>
      <td><code>500</code></td>
      <td><code>Integer</code> How long the scroller will wait after the resizing to initiate recalculation
      (this is done to ensure that the recalculation takes place only after the user has finished resizing the screen).
      </td>
    </tr>
    <tr>
      <th scope="row"><code>easingType</code></th>
      <td><code>"swing"</code></td>
      <td><code>String</code> Easing type for the scrolling. You can specify any other type of easing but bear in mind
      that jQuery has only <code>swing</code> and <code>linear</code> easings included.
      You will have to include
      <a target="_blank" href="http://jqueryui.com/effect/#easing">jQueryUI's effects and easings plugin</a> to get more.</td>
    </tr>
    <tr>
      <th scope="row"><code>elementMargin</code></th>
      <td><code>10</code></td>
      <td><code>Integer</code> Horizontal (left/right) margin for the scroller's elements. This value should correspond
      to the value that is provided in the styles.</td>
    </tr>
    <tr>
      <th scope="row"><code>elementWidth</code></th>
      <td><code>200</code></td>
      <td><code>Integer</code> Width for the scroller's elements. This value should correspond
      to the value that is provided in the styles.</td>
    </tr>
    <tr>
      <th scope="row"><code>nextButtonText</code></th>
      <td><code>""</code></td>
      <td><code>String</code> Text to be shown on the "next" button.</td>
    </tr>
    <tr>
      <th scope="row"><code>prevButtonText</code></th>
      <td><code>""</code></td>
      <td><code>String</code> Text to be shown on the "previous" button.</td>
    </tr>
    <tr>
      <th scope="row"><code>rows</code></th>
      <td><code>2</code></td>
      <td><code>Integer</code> How many rows should the scroller have.</td>
    </tr>
    <tr>
      <th scope="row"><code>useCaptions</code></th>
      <td><code>false</code></td>
      <td><code>Boolean</code> Whether captions should be enabled. Check <a href="#captions">Captions</a>
      section for more information.</td>
    </tr>
    <tr>
      <th scope="row"><code>useMouseWheel</code></th>
      <td><code>false</code></td>
      <td><code>Boolean</code> Whether support for the mousewheel scrolling should be enabled. Please note
      that when this value is set to <code>true</code>, you should hook up
      <a target="_blank" href="https://github.com/brandonaaron/jquery-mousewheel">MouseWheel plugin</a>.</td>
    </tr>
  </tbody>
</table>

## Styling

`jquery.radiant_scroller.css` provides only required styles so you have to write some more styles to make the scroller
look nice. When the scroller is initialized the following layout is being built (pagination will not be shown until
you set `addPagination` option to `true`):

```html
<div class="radiant_scroller">
    <div class="radiant_scroller_wrapper">
        <div id="myScroller">
            <div class="radiant_scroller_row">
                <div class="scroller-el"><img src="image1.jpg" alt="Image1" /></div>
                <div class="scroller-el"><img src="image2.jpg" alt="Image2" /></div>
            </div>
            <div class="radiant_scroller_row">
                <div class="scroller-el"><img src="image3.jpg" alt="Image3" /></div>
                <div class="scroller-el"><img src="image4.jpg" alt="Image4" /></div>
            </div>
        </div>
    </div>
    <div class="radiant-pagination">
        <div class="radiant-page current-page" data-page="1"></div>
        <div class="radiant-page" data-page="2"></div>
    </div>
    <div class="radiant-navigation">
        <div class="radiant-prev"></div>
        <div class="radiant-next"></div>
    </div>
</div>
```

All elements except for `#myScroller` and `.scroller-el` are added dynamically.

* `.radiant-pagination` contains navigational pages; the active page has the `current-page` class, so you can style
it differently.
* `.radiant-navigation` contains next and previous buttons. `.radiant-prev` and `.radiant-next` are positioned absolutely
(and main wrapper `.radiant_scroller` is positioned relatively) so you can adjust its position as necessary.

See [demos](#demos) to get the basic idea how the scroller can be styled.

## Captions

Starting from version 0.1.0 scroller elements may have animated captions. To use them few things have to be done:

* Set `useCaptions` option to `true`
* Adjust `captionsAnimateDuration` and `captionsAnimateEasingType` if needed
* Add `title` attribute to your elements like this:

```html
<div id="your_scroller_id">
    <div class="scroller-el"><img src="image1.jpg" alt="Image1" title="This is a caption" /></div>
    <div class="scroller-el"><img src="image2.jpg" alt="Image2" title="Another caption" /></div>
</div>
```

* Enjoy the captions!

By default captions appear at the bottom of the element with white centered text and black semi-opaque background. If
you wish
to style it differently, assign your styles to `.radiant-caption` class.

## API

To get access to the RadiantScroller's API you should initialize your scroller like this:

```js
var my_scroller = $('#myScroller').radiantScroller({...});
```

And then you can manage your scroller by calling `radiantScroller` on the `my_scroller` variable and passing it an
API action to invoke. Currently there are a few API methods available:

* `radiantScroller('next')` - scroll one page forward
* `radiantScroller('prev')` - scroll one page backward
* `radiantScroller(<number>)` - scroll to a page with the specified number. Page numeration starts from 1. If a
non-existent page is provided nothing happens.
* `radiantScroller('by', <number>)` - scroll by a specified number of pages. For example if you are at the 1st page and
call `my_scroller.radiantScroller('by', 2)` you scroll by 2 pages and end up at the 3rd page.

To see them all in action open [this demo](http://www.radiant-wind.com/code_examples/13).
More methods coming soon.

## Callbacks

Starting from version 0.1.0 the following callbacks are available:

<table>
  <thead>
    <tr>
      <th>Callback</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th scope="row"><code>loaded</code></th>
      <td>Fires when the scroller has finished loading and is ready. Scroller is being passed as an argument.</td>
    </tr>
    <tr>
      <th scope="row"><code>beforeMove</code></th>
      <td>Fires before each scroller's moving animation. Scroller is being passed as an argument.</td>
    </tr>
    <tr>
      <th scope="row"><code>afterMove</code></th>
      <td>Fires before after scroller's moving animation. Scroller is being passed as an argument.</td>
    </tr>
    <tr>
      <th scope="row"><code>lastPageReached</code></th>
      <td>Fires when the scroller has finished loading and is ready. Scroller is being passed as an argument.</td>
    </tr>
    <tr>
      <th scope="row"><code>afterHidingCaption</code></th>
      <td>Fires after the caption was hidden. Caption is being passed as an argument.</td>
    </tr>
    <tr>
      <th scope="row"><code>afterShowingCaption</code></th>
      <td>Fires after the caption was showed. Caption is being passed as an argument.</td>
    </tr>
  </tbody>
</table>

## For Rails developers

[jquery-radiantscroller-rails](https://github.com/bodrovis/jquery-radiantscroller-rails) gem provides an easy way
to integrate Radiant Scroller into your Rails project.

## License

This plugin is licensed under the [MIT License](https://github.com/bodrovis/RadiantScroller/blob/master/LICENSE.txt).

Copyright (c) 2014 [Ilya Bodrov](http://radiant-wind.com)


CHANGELOG.md
# RadiantScroller changelog

## 0.1.0 (29/07/2014)

* Elements now may have captions
* Added `useCaptions`, `captionsAnimateDuration` and `captionsAnimateEasingType` options
* Added callback API

## 0.0.5 (28/07/2014)

* Fixed an error caused to crop some elements during resizing
* Added `delayDuration` option

## 0.0.4 (23/07/2014)

* Fixed hanging of the page when resized to very low width

## 0.0.3 (08/01/2014)

* Reworked pagination and scrolling
* Added `radiantScroller(<number>)` API method
* Added `radiantScroller('by', <number>)` API method

## 0.0.2 (07/01/2014)

* Docs and links

## 0.0.1 (07/01/2014)

* Basic functionality.
* Next/previos API methods.

jquery.radiant_scroller-min.js
/*!
 * jQuery RadiantScroller
 * Version: 0.0.2 (07/01/2014)
 * Copyright (c) 2013 Ilya Bodrov (http://www.radiant-wind.com/plugins/radiant_scroller)
 *
 * Requires: jQuery 1.7.0+
 */
(function(e){e.radiantScroller=function(t,n){var r=e(t);r.vars=e.extend({},e.radiantScroller.defaults,n);r.data("radiantscroller",r);r.animating=false;var i=r.find(".scroller-el"),s=i.size(),o=r.vars.elementWidth+r.vars.elementMargin,u=Math.ceil(s/r.vars.rows),a=0;r.width(o*u-r.vars.elementMargin);do e(i.slice(0,u)).wrapAll('<div class="radiant_scroller_row" />');while((i=i.slice(u)).length>0);r.wrap('<div class="radiant_scroller"></div>').wrap('<div class="radiant_scroller_wrapper" />');var f=r.parent(".radiant_scroller_wrapper");var l=f.parent(".radiant_scroller");var c=r.vars.cols*o-r.vars.elementMargin+"px";f.css("max-width",c);l.css("max-width",c);var h=e('<div class="radiant-navigation" />').insertAfter(f);var p=h.append(e('<div class="radiant-prev">'+r.vars.prevButtonText+"</div>"));var d=h.append(e('<div class="radiant-next">'+r.vars.nextButtonText+"</div>"));r.calculateVisibleElements=function(){a=Math.floor((f.width()+r.vars.elementMargin)/o)};r.moveElements=function(e,t,n){if(!r.animating){r.animating=true;var i=0;if(typeof t==="undefined"){i=a*o}else{i=a*o*t}var s=0;if(e==="left"){s=-i}else{s=i}f.animate({scrollLeft:f.scrollLeft()+s},r.vars.animateDuration,r.vars.easingType,function(){r.animating=false});if(r.vars.addPagination){f.current_page.removeClass("current-page");if(typeof n==="undefined"){if(e==="right"){if(f.current_page.next().size()>0)f.current_page=f.current_page.next()}else{if(f.current_page.prev().size()>0)f.current_page=f.current_page.prev()}}else{f.current_page=n}f.current_page.addClass("current-page")}}};r.createPagination=function(){if(typeof this.visible_els==="undefined"||this.visible_els!==a){this.visible_els=a;var t=Math.ceil(u/a);if(l.find(".radiant-pagination").size()>0)l.find(".radiant-pagination").remove();var n=e('<div class="radiant-pagination" />').insertAfter(f);for(var r=0;r<t;r++){n.append('<div class="radiant-page" data-page="'+(r+1)+'"></div>')}if(typeof f.current_page==="undefined"){f.current_page=n.find(".radiant-page").first()}else{var i=Math.ceil(f.scrollLeft()/(a*o));f.current_page=e(n.find(".radiant-page").get(i))}f.current_page.addClass("current-page")}};e(window).bindWithDelay("resize",function(){r.calculateVisibleElements();if(r.vars.addPagination){r.createPagination()}},500,true);h.on("click",".radiant-next",function(){r.moveElements("right",1)});h.on("click",".radiant-prev",function(){r.moveElements("left",1)});if(r.vars.useMouseWheel){f.on("mousewheel",function(e){e.preventDefault();if(e.deltaY>0){r.moveElements("right")}else{r.moveElements("left")}})}l.on("click",".radiant-page",function(){var t=e(this);if(f.current_page.get(0)!==t.get(0)){var n=f.current_page.data("page");var i=t.data("page");if(n>i){r.moveElements("left",n-i,t)}else{r.moveElements("right",i-n,t)}}});r.calculateVisibleElements();if(r.vars.addPagination){r.createPagination()}};e.radiantScroller.defaults={cols:2,elementWidth:200,elementMargin:10,easingType:"swing",animateDuration:700,rows:2,useMouseWheel:false,addPagination:false,nextButtonText:"",prevButtonText:""};e.fn.radiantScroller=function(t){if(t===undefined)t={};if(typeof t==="object"){new e.radiantScroller(this,t)}else{var n=e(this).data("radiantscroller");switch(t){case"next":n.moveElements("right");break;case"prev":n.moveElements("left");break}}return this}})(jQuery);(function(e){e.fn.bindWithDelay=function(t,n,r,i,s){if(e.isFunction(n)){s=i;i=r;r=n;n=undefined}r.guid=r.guid||e.guid&&e.guid++;return this.each(function(){function u(){var t=e.extend(true,{},arguments[0]);var n=this;var u=function(){o=null;r.apply(n,[t])};if(!s){clearTimeout(o);o=null}if(!o){o=setTimeout(u,i)}}var o=null;u.guid=r.guid;e(this).bind(t,n,u)})}})(jQuery)

jquery.radiant_scroller.js
/*!
 * jQuery RadiantScroller
 * Version: 0.0.4 (23/07/2014)
 * Copyright (c) 2014 Ilya Bodrov (http://www.radiant-wind.com/plugins/radiant_scroller)
 *
 * Requires: jQuery 1.7.0+
 */

(function($) {
    $.radiantScroller = function(el, options) {
        var scroller = $(el);

        scroller.vars = $.extend({}, $.radiantScroller.defaults, options);

        // Store reference to scroller in data of the object
        scroller.data("radiantscroller", scroller);
        // Set to true when scrolling
        scroller.animating = false;

        var elements = scroller.find('.scroller-el'),
            el_count = elements.size(),
            el_width = scroller.vars.elementWidth + scroller.vars.elementMargin,
            per_row = Math.ceil(el_count / scroller.vars.rows);

        // Calculating scroller's width with regard to items per row
        scroller.width( (el_width * per_row) - scroller.vars.elementMargin );

        // Grouping elements by rows
        do $(elements.slice(0, per_row)).wrapAll('<div class="radiant_scroller_row" />');
        while((elements = elements.slice(per_row)).length > 0);

        // Wrapper with hidden scrollbars
        scroller.wrap('<div class="radiant_scroller"></div>').wrap('<div class="radiant_scroller_wrapper" />');
        var wrapper = scroller.parent('.radiant_scroller_wrapper');
        var outer_wrapper = wrapper.parent('.radiant_scroller');

        var max_wrapper_width = (scroller.vars.cols * el_width - scroller.vars.elementMargin) + 'px';
        wrapper.css('max-width', max_wrapper_width);
        outer_wrapper.css('max-width', max_wrapper_width);

        // Next/previous buttons
        var nav = $('<div class="radiant-navigation" />').insertAfter(wrapper);
        var prev = nav.append($('<div class="radiant-prev">' + scroller.vars.prevButtonText + '</div>'));
        var next = nav.append($('<div class="radiant-next">' + scroller.vars.nextButtonText + '</div>'));

        // Scroller methods
        scroller.calculateVisibleElements = function() {
            scroller.visible_els = (wrapper.width() + scroller.vars.elementMargin) / el_width;
        };

        scroller.moveElements = function(scrollBy) {
            if (!scroller.animating) {
                if (typeof scrollBy === 'undefined')
                    scrollBy = 1;

                var new_current_page = scroller.current_page + scrollBy;
                // The requested page should be between 0 and total number of pages and should not be the same page
                // as the current, otherwise do nothing
                if (new_current_page >= 0 && new_current_page < scroller.total_pages &&
                    new_current_page !== scroller.current_page) {
                    scroller.animating = true;

                    wrapper.animate(
                        { scrollLeft: wrapper.scrollLeft() + scroller.visible_els * el_width * scrollBy },
                        scroller.vars.animateDuration, scroller.vars.easingType,
                        function() { scroller.animating = false; }
                    );

                    scroller.current_page = new_current_page;

                    if (scroller.vars.addPagination)
                        scroller.assignActivePage(true);
                }
            }
        };

        scroller.initializePagination = function() {
            // If there is the same number of visible elements we don't have to change anything
            if (typeof scroller.old_visible_els === 'undefined' || scroller.old_visible_els !== scroller.visible_els) {
                scroller.old_visible_els = scroller.visible_els;
                scroller.total_pages = Math.ceil(per_row / scroller.visible_els);

                if (typeof scroller.current_page === 'undefined') {
                    // If current page was not set previously, we set it now
                    scroller.current_page = 0;
                } else {
                    scroller.current_page = Math.ceil(wrapper.scrollLeft() / (scroller.visible_els * el_width));
                }

                if (scroller.vars.addPagination)
                    scroller.drawPagination();
            }
        };

        scroller.drawPagination = function() {
            if (outer_wrapper.find('.radiant-pagination').size() > 0)
                outer_wrapper.find('.radiant-pagination').remove();

            var pagination = $('<div class="radiant-pagination" />').insertAfter(wrapper);
            for (var i = 0; i < scroller.total_pages; i++) {
                pagination.append('<div class="radiant-page" data-page="' + i + '" />');
            }
            scroller.assignActivePage();
        };

        scroller.assignActivePage = function(remove_old_active) {
            if (remove_old_active === true)
                outer_wrapper.find('.current-page').removeClass('current-page');
            $(outer_wrapper.find('.radiant-page').get(scroller.current_page)).addClass('current-page');
        };

        // Binding events
        $(window).bindWithDelay('resize', function() {
            scroller.calculateVisibleElements();
            scroller.initializePagination();
        }, 500, true);

        nav.on('click', '.radiant-next', function() {
            scroller.moveElements(1);
        });

        nav.on('click', '.radiant-prev', function() {
            scroller.moveElements(-1);
        });

        if (scroller.vars.useMouseWheel) {
            wrapper.on('mousewheel', function(event) {
                event.preventDefault();
                scroller.moveElements(event.deltaY);
            });
        }

        outer_wrapper.on('click', '.radiant-page', function() {
            var this_page = $(this).data('page');
            // If not clicking on the same page
            if (scroller.current_page !== this_page) {
                var current_page_number = scroller.current_page;
                scroller.moveElements(this_page - current_page_number);
            }
        });

        // Init scroller
        scroller.calculateVisibleElements();
        scroller.initializePagination();
    };

    $.radiantScroller.defaults = {
        cols: 2,
        elementWidth: 200,
        elementMargin: 10,
        easingType: 'swing',
        animateDuration: 700,
        rows: 2,
        useMouseWheel: false,
        addPagination: false,
        nextButtonText: '',
        prevButtonText: ''
    };

    $.fn.radiantScroller = function(options, options2) {
        if (options === undefined) options = {};

        if (typeof options === "object") {
            new $.radiantScroller(this, options);
        } else {
            var $scroller = $(this).data('radiantscroller');
            switch (options) {
                case "next": $scroller.moveElements(1); break;
                case "prev": $scroller.moveElements(-1); break;
                // radiantScroller('by', 2) - scroll by 2 pages, not to the 2nd page
                //TODO: check number
                case "by": if (typeof options2 === "number") $scroller.moveElements(options2); break;
                // Default behaviour: radiantScroller(3) means go to 3 page and page numeration starts from 1
                // (whereas plugin has numeration starting from 0, so have to substract 1!!!)
                default: if (typeof options === "number") $scroller.moveElements(options - 1 - $scroller.current_page);
            }
        }

        return this;
    };
})(jQuery);

/*
 bindWithDelay jQuery plugin
 Author: Brian Grinstead
 MIT license: http://www.opensource.org/licenses/mit-license.php

 http://github.com/bgrins/bindWithDelay
 http://briangrinstead.com/files/bindWithDelay

 Usage:
 See http://api.jquery.com/bind/
 .bindWithDelay( eventType, [ eventData ], handler(eventObject), timeout, throttle )

 Examples:
 $("#foo").bindWithDelay("click", function(e) { }, 100);
 $(window).bindWithDelay("resize", { optional: "eventData" }, callback, 1000);
 $(window).bindWithDelay("resize", callback, 1000, true);
 */

(function($) {

    $.fn.bindWithDelay = function( type, data, fn, timeout, throttle ) {

        if ( $.isFunction( data ) ) {
            throttle = timeout;
            timeout = fn;
            fn = data;
            data = undefined;
        }

        // Allow delayed function to be removed with fn in unbind function
        fn.guid = fn.guid || ($.guid && $.guid++);

        // Bind each separately so that each element has its own delay
        return this.each(function() {

            var wait = null;

            function cb() {
                var e = $.extend(true, { }, arguments[0]);
                var ctx = this;
                var throttler = function() {
                    wait = null;
                    fn.apply(ctx, [e]);
                };

                if (!throttle) { clearTimeout(wait); wait = null; }
                if (!wait) { wait = setTimeout(throttler, timeout); }
            }

            cb.guid = fn.guid;

            $(this).bind(type, data, cb);
        });
    };

})(jQuery);